Landing pages table new columns

alfredoct96 opened the pull request:
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | features
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | N/A

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "features" branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:

Added three new columns to landing pages table:
* created at
* created by
* modified at

Changed default order to decreasing modified at.

### Before
![landingpagesbefore](https://user-images.githubusercontent.com/50916237/109321442-7117c780-7851-11eb-86de-453e62f884dc.png)

### After
![landingafter](https://user-images.githubusercontent.com/50916237/109321545-8b51a580-7851-11eb-9b3a-606a36fa72e2.png)



<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create two or more landing pages
3. Check existence of new columns
4. Check default order

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9691"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

alfredoct96 committed:
Added new sortable columns to landing pages table, created at, modified date, createdby

alfredoct96 committed:
Set default order to decreasing modified date

alfredoct96 committed:
At creation of landing set also modified date

alfredoct96 committed:
Added index action test to PageControllerTest

alfredoct96 committed:
changed format date

dennisameling commented:
Hi @alfredoct96! Thanks for your PR 🚀 we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since we don't have access to your branch, could you please click "Update branch" at the bottom of this PR? If that doesn't work, you can also run `git merge upstream/features` locally. Thanks in advance!

![image](https://user-images.githubusercontent.com/17739158/112621033-90dae500-8e29-11eb-8541-bd0a5d36c2ba.png)

alfredoct96 commented:
> Hi @alfredoct96! Thanks for your PR  we just upgraded Mautic to Symfony 4 and this PR will need to be updated. Since we don't have access to your branch, could you please click "Update branch" at the bottom of this PR? If that doesn't work, you can also run `git merge upstream/features` locally. Thanks in advance!
> 
> ![image](https://user-images.githubusercontent.com/17739158/112621033-90dae500-8e29-11eb-8541-bd0a5d36c2ba.png)

Done @dennisameling

kuzmany reviewed (changes requested):
Works.
But could we change same date formatting like for segments ?

![image](https://user-images.githubusercontent.com/462477/153570707-be42bcaa-9e39-4156-9565-694847fbe509.png)

alfredoct96 committed:
refactor(pagelist): use long date on tooltip

alfredoct96 commented:
> Works. But could we change same date formatting like for segments ?
> 
> ![image](https://user-images.githubusercontent.com/462477/153570707-be42bcaa-9e39-4156-9565-694847fbe509.png)

Done @kuzmany

kuzmany commented on app/bundles/PageBundle/Tests/Controller/PageControllerTest.php:
```diff
@@ -45,6 +45,16 @@ protected function setUp(): void
+     */
+    public function testIndexAction(): void
+    {
+        $this->client->request('GET', '/s/pages');
```
These kind of tests unfinished
You should check If you Landing pages page should return date created/modified column created by, and if modified date header contain icon with class `fa-sort-amount-desc`

kuzmany commented on app/bundles/PageBundle/Views/Page/list.html.php:
```diff
@@ -62,6 +61,37 @@
+                        'orderBy'    => 'p.dateModified',
+                        'text'       => 'mautic.lead.import.label.dateModified',
+                        'class'      => 'col-page-dateModified visible-md visible-lg',
+                        'default'    => true,
```
Here you need define another  attribute:
`dir => desc`

By default icon is set to ASC. I think it's related to all your PRs with columns

![image](https://user-images.githubusercontent.com/462477/153743736-a67f4008-e9ed-42de-87ec-311723c53c67.png)

alfredoct96 replied:
I can't set `dir` to desc because if i do it icon will be fixed to DESC, i added another parameter `orderDir`  that will be considered if `default` param is true.
After this fix if you click a second time on modified date the order will be continue to be DESC, this happens setListFilters method on CommonController has ASC as default behavior, it doesn't see orderbydir variable setted in session and convert again to DESC. So i needed to do another fix : set session orderbydir to DESC inside PageController if it's not defined.

kuzmany replied:
You're right.  But I think it could be done more simple.

What about just new param defaultDir => 'DESC' what replace 

`$dir          = (!empty($dir)) ? $dir : $app->getSession()->get("mautic.{$sessionVar}.orderbydir", $defaultDir ?? 'ASC');`

alfredoct96 replied:
Ok i changed it, i also added new tests

alfredoct96 committed:
fix(tableorder): at default order icon is asc and not desc, i change the default behavior

Changed default icon to asc and added another option "orderDir" for default icon, cant be used
option "dir" because it will be fixed the icon at one order. I also set orderbydir session at init
in PageController, because setListFilters method inside CommonController has ASC as default
behavior. If don't do it after first click on modified date icon the order will be again DESC, i
need that it will be converted to ASC.

kuzmany commented on app/bundles/PageBundle/Controller/PageController.php:
```diff
@@ -113,8 +113,14 @@ public function indexAction($page = 1)
-        $orderByDir = $this->get('session')->get('mautic.page.orderbydir', 'DESC');
+        $orderBy    = $this->get('session')->get('mautic.page.orderby', 'p.dateModified');
+
+        $orderByDir = $this->get('session')->get('mautic.page.orderbydir');
```
We can replace it with one line:

`$orderByDir = $this->get('session')->get('mautic.page.orderbydir', 'DESC')`

alfredoct96 committed:
refactor(pageslist): changed init of dir param

alfredoct96 committed:
test(pagecontrollertest): added tests for page list, check if columns exist and order is correct

kuzmany commented on app/bundles/PageBundle/Controller/PageController.php:
```diff
@@ -113,8 +113,14 @@ public function indexAction($page = 1)
-        $orderByDir = $this->get('session')->get('mautic.page.orderbydir', 'DESC');
+        $orderBy    = $this->get('session')->get('mautic.page.orderby', 'p.dateModified');
+
+        $orderByDir = $this->get('session')->get('mautic.page.orderbydir');
+        if (!$orderByDir) {
+            $orderByDir = 'DESC';
+            $this->get('session')->set('mautic.page.orderbydir', $orderByDir);
+        }
```
These lines do not need, do we?

alfredoct96 replied:
we need these lines because if we don't set session to DESC and we click on modified date order the order will be remains to DESC, this because  setListFilters of CommonController set order to DESC if order doesn't exists in session.

These are the lines of code:
``` 
 if ($this->request->query->has('orderby')) {
       $orderBy = InputHelper::clean($this->request->query->get('orderby'), true);
       $dir     = $session->get("$name.orderbydir", 'ASC');
       $dir     = ('ASC' == $dir) ? 'DESC' : 'ASC';
       $session->set("$name.orderby", $orderBy);
       $session->set("$name.orderbydir", $dir);
   }
```

kuzmany replied:
I did test without these lines and it works.
I would like push the changes, but it return permission issue.

alfredoct96 replied:
Have you tried to enter go into the list and click on modify date order? However i pushed the changes

kuzmany replied:
I have solution for it.  

Can you merge actual 4.x branch, because it wrote:

> This branch is out-of-date with the base branch

alfredoct96 replied:
Done

alfredoct96 replied:
I saw the solution, it seems a good solution

kuzmany replied:
I still see 

> This branch is out-of-date with the base branch

alfredoct96 replied:
Done again

alfredoct96 committed:
refactor(pagecontroller): remove if condition and use default order to session
